### PR TITLE
Retain process logs on robustness tests

### PR DESCRIPTION
**Pull request by tjungblu**
Adding a new flag to retain e2e etcd process logs after stop and saving next to the visualized model.

Spun out of #16018 where I used it for easier local debugging on model violations.

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by serathius** [comment]
Etcd logs are available the same way as e2e logs. By passing `EXPECT_DEBUG=true` flag to the test run.

**Comment by tjungblu** [comment]
> Etcd logs are available the same way as e2e logs. By passing EXPECT_DEBUG=true flag to the test run.

If you feel like browsing the logs is convenient enough already I'm happy to close. For me it makes more sense to have the member debug logs next to the history.html in the artifacts zip.

File: tests/framework/e2e/etcd_process.go [hunk 1]
```diff
@@ -57,6 +57,7 @@ type EtcdProcess interface {
 	PeerProxy() proxy.Server
 	Failpoints() *BinaryFailpoints
 	Logs() LogsExpect
+	LogLines() []string
```
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:60]: Please add Lines() to `LogsExpect` interface

**Comment by serathius** [comment]
> If you feel like browsing the logs is convenient enough already I'm happy to close. For me it makes more sense to have the member debug logs next to the history.html in the artifacts zip.

It's not https://github.com/etcd-io/etcd/issues/15079

File: tests/framework/e2e/etcd_process.go [hunk 2]
```diff
@@ -72,6 +73,8 @@ type EtcdServerProcess struct {
 	proxy      proxy.Server
 	failpoints *BinaryFailpoints
 	donec      chan struct{} // closed when Interact() terminates
+
+	keptLogLines []string // only filled when KeepLogs=true
```
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:77]: Not sure why this is needed. ExpectLogs already stores all logs in memory. Don't see a problem to always keep logs.

Comment (tjungblu) [reply tests/framework/e2e/etcd_process.go:77]: it's because of the ep.proc lifecycle below:

```
		if ep.cfg.KeepLogs {
			ep.keptLogLines = ep.proc.Lines()
		}
***		ep.proc = nil ***
```

it's also used to flag whether a process was stopped, which coincidentally wipes the logs with it.

Comment (serathius) [reply tests/framework/e2e/etcd_process.go:77]: Then I would suggest storing `ExpectLog` interface instead of `[]string{}`

Comment (tjungblu) [reply tests/framework/e2e/etcd_process.go:77]: sounds good

**Comment by serathius** [comment]
FYI main challenge here is to merge logs from between restarts.

**Comment by tjungblu** [comment]
@serathius created an aggregation of LogExpects to merge the logs. Let me know if the implemented behavior works for you, too (mostly on the expect side of things).

File: tests/framework/e2e/etcd_process.go [hunk 3]
```diff
@@ -230,7 +233,7 @@ func (ep *EtcdServerProcess) Config() *EtcdServerProcessConfig { return ep.cfg }
 
 func (ep *EtcdServerProcess) Logs() LogsExpect {
```
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:234]: Code looks nice, but changing `func (ep *EtcdServerProcess) Logs() LogsExpect` semantic behavior is risky. When I expect most people will assume `member.Logs().ExpectWithContext()` will analyse logs emitted from the current run and might stumble on problem that log they are expecting was already emitted before restart. 

Proposal:
* Have `func (ep *EtcdServerProcess) Logs() LogsExpect` return logsExpect from the latest run `ep.logs.procLogs[len(ep.logs.procLogs)-1]`
* Have a separate method like `allLogLines` or `aggregatedLogLines` etcd that will return appended results of all runs.

What do you think?

Comment (chaochn47) [reply tests/framework/e2e/etcd_process.go:234]: Agreed with the incremental changes to be safe. 

`func (ep *EtcdServerProcess) Logs() LogsExpect` was using `ep.proc` in happy case and most of the cases. 

The PR is intended to retain all the previous runs logs after all the runs of the process are stopped. 

Seems only `func (m *multiLogsExpect) Lines()` is needed.

Comment (tjungblu) [reply tests/framework/e2e/etcd_process.go:234]: > Have func (ep *EtcdServerProcess) Logs() LogsExpect return logsExpect from the latest run ep.logs.procLogs[len(ep.logs.procLogs)-1]

Sounds like a good proposal to me. Is that also acceptable for you @chaochn47?

> Seems only func (m *multiLogsExpect) Lines() is needed.

Currently yes, I'm surprised we don't have expect tests across stop/restart scenarios.

Comment (serathius) [reply tests/framework/e2e/etcd_process.go:235]: > Currently yes, I'm surprised we don't have expect tests across stop/restart scenarios.

We want to explicitly avoid that because it not natural. Imagine that you want to check that etcd member has started using one of the starting logs. It's natural to call `member.Logs().Expect("etcd started")`, but it needs to only consider the logs from the last restart to make sense. Otherwise it might always return true when member was already restarted.

tldr. Don't want to change the semantics of `func (ep *EtcdServerProcess) Logs() LogsExpect`.

Comment (tjungblu) [reply tests/framework/e2e/etcd_process.go]: Cool, pushed another by just moving up the Lines and Count via interface, so the current LogsExpect stays entirely untouched.

File: tests/framework/e2e/etcd_process.go [hunk 4]
```diff
@@ -235,6 +242,14 @@ func (ep *EtcdServerProcess) Logs() LogsExpect {
 	return ep.proc
 }
 
+func (ep *EtcdServerProcess) Lines() []string {
+	return ep.logs.Lines()
+}
+
+func (ep *EtcdServerProcess) LineCount() int {
```
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:249]: Doesn't looks like it's used by anyone. Remove?

File: tests/framework/e2e/etcd_process.go [hunk 5]
```diff
@@ -72,6 +77,7 @@ type EtcdServerProcess struct {
 	proxy      proxy.Server
 	failpoints *BinaryFailpoints
 	donec      chan struct{} // closed when Interact() terminates
+	logs       *procLogs
```
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:80]:
```suggestion
	logs       procLogs
```
It doesn't need to be a pointer you know?

File: tests/framework/e2e/etcd_process.go (hunk 4 above)
Comment (serathius) [reply tests/framework/e2e/etcd_process.go:249]: I mean it's not needed for EtcdServerProcess.

**Comment by serathius** [comment]
Also, please remove environment EXPECT_DEBUG from the Robustness tests github workflow.
